diff -u a/dhcpcd-hooks/Makefile b/dhcpcd-hooks/Makefile
--- a/dhcpcd-hooks/Makefile	2015-09-18 14:47:00.345622850 -0700
+++ b/dhcpcd-hooks/Makefile	2015-09-18 14:47:38.970840854 -0700
@@ -4,7 +4,7 @@

 SCRIPTSDIR=	${LIBEXECDIR}/dhcpcd-hooks
 SCRIPTS=	01-test 02-dump
-SCRIPTS+=	10-mtu 20-resolv.conf 29-lookup-hostname 30-hostname
+SCRIPTS+=	10-mtu 20-resolv.conf 29-lookup-hostname 30-hostname 50-dhcpcd-compat
 SCRIPTS+=	${HOOKSCRIPTS}

 all:
