From 6f26d5b0ae49911ce7bb3c7f9621064099a6cbee Mon Sep 17 00:00:00 2001
From: Viswanath Kraleti <vkraleti@codeaurora.org>
Date: Mon, 2 Sep 2013 16:01:59 +0530
Subject: [PATCH] ICU: Disable renaming APIs.

Change-Id: Id9a796be3d3b6ec96f27bfdd6401f4e75d55cc07
---

--- a/common/unicode/urename.h
+++ b/common/unicode/urename.h
@@ -23,7 +23,7 @@
 
 /* Uncomment the following line to disable renaming on platforms
    that do not use Autoconf. */
-/* #define U_DISABLE_RENAMING 1 */
+#define U_DISABLE_RENAMING 1
 
 #if !U_DISABLE_RENAMING
 
-- 
1.8.2.1

