diff --git a/arch/arm/mach-msm/board-msm7627a.h b/arch/arm/mach-msm/board-msm7627a.h
index e3c9469..ff48fa7 100644
--- a/arch/arm/mach-msm/board-msm7627a.h
+++ b/arch/arm/mach-msm/board-msm7627a.h
@@ -21,7 +21,7 @@ void __init msm7627a_init_mmc(void);
 void __init msm_msm7627a_allocate_memory_regions(void);
 void __init msm_fb_add_devices(void);
 
-#define PMEM_KERNEL_EBI1_SIZE	0x3A000
+#define PMEM_KERNEL_EBI1_SIZE	0x1200000
 #define MSM_PMEM_AUDIO_SIZE	0x1F4000
 
 #define MSM_PMEM_MDP_SIZE		0x2300000
