diff --git a/QCamera2/stack/mm-camera-interface/src/mm_camera_interface.c b/QCamera2/stack/mm-camera-interface/src/mm_camera_interface.c
index cdd9ab9..1311cbd 100644
--- a/QCamera2/stack/mm-camera-interface/src/mm_camera_interface.c
+++ b/QCamera2/stack/mm-camera-interface/src/mm_camera_interface.c
@@ -1977,13 +1977,11 @@ static int32_t mm_camera_intf_register_stream_buf_cb(uint32_t camera_handle,
     }
     return (int32_t)rc;
 }
-
 struct camera_info *get_cam_info(uint32_t camera_id, cam_sync_type_t *pCamType)
 {
-    *pCamType = g_cam_ctrl.cam_type[camera_id];
+    *pCamType = CAM_TYPE_STANDALONE;
     return &g_cam_ctrl.info[camera_id];
 }
-
 uint8_t is_yuv_sensor(uint32_t camera_id)
 {
     return g_cam_ctrl.is_yuv[camera_id];
