From e73c18ffe48ed37938b9dbff71f354b8d414c326 Mon Sep 17 00:00:00 2001
From: Tyler Wear <twear@codeaurora.org>
Date: Mon, 8 Jul 2013 14:17:02 -0700
Subject: [PATCH 1/1] mpeg-codec-fix

---
 metadata.c |  212 ++++++++++++++++++++++++++++--------------------------------
 1 files changed, 100 insertions(+), 112 deletions(-)
 mode change 100644 => 100755 metadata.c

diff --git a/metadata.c b/metadata.c
old mode 100644
new mode 100755
index 3e5f4db..e724b2f
--- a/metadata.c
+++ b/metadata.c
@@ -1176,133 +1176,80 @@ GetVideoMetadata(const char * path, char * name)
				{
					off += sprintf(m.dlna_pn+off, "MP4_");

-					switch( vc->profile ) {
-					case FF_PROFILE_H264_BASELINE:
-						if( vc->width  <= 352 &&
-						    vc->height <= 288 )
-						{
-							if( ctx->bit_rate < 600000 )
-								off += sprintf(m.dlna_pn+off, "BL_CIF15_");
-							else if( ctx->bit_rate < 5000000 )
-								off += sprintf(m.dlna_pn+off, "BL_CIF30_");
-							else
-								goto mp4_mp_fallback;
+					if( vc->width  <= 352 &&
+						vc->height <= 288 )
+					{
+						if( ctx->bit_rate < 600000 )
+							off += sprintf(m.dlna_pn+off, "BL_CIF15_");
+						else if( ctx->bit_rate < 5000000 )
+							off += sprintf(m.dlna_pn+off, "BL_CIF30_");
+						else
+							goto mp4_mp_fallback;

-							if( audio_profile == PROFILE_AUDIO_AMR )
+						if( audio_profile == PROFILE_AUDIO_AMR )
+						{
+							off += sprintf(m.dlna_pn+off, "AMR");
+						}
+						else if( audio_profile == PROFILE_AUDIO_AAC )
+						{
+							off += sprintf(m.dlna_pn+off, "AAC_");
+							if( ctx->bit_rate < 520000 )
							{
-								off += sprintf(m.dlna_pn+off, "AMR");
+								off += sprintf(m.dlna_pn+off, "520");
							}
-							else if( audio_profile == PROFILE_AUDIO_AAC )
+							else if( ctx->bit_rate < 940000 )
							{
-								off += sprintf(m.dlna_pn+off, "AAC_");
-								if( ctx->bit_rate < 520000 )
-								{
-									off += sprintf(m.dlna_pn+off, "520");
-								}
-								else if( ctx->bit_rate < 940000 )
-								{
-									off += sprintf(m.dlna_pn+off, "940");
-								}
-								else
-								{
-									off -= 13;
-									goto mp4_mp_fallback;
-								}
+								off += sprintf(m.dlna_pn+off, "940");
							}
							else
							{
-								off -= 9;
+								off -= 13;
								goto mp4_mp_fallback;
							}
						}
-						else if( vc->width  <= 720 &&
-						         vc->height <= 576 )
-						{
-							if( vc->level == 30 &&
-							    audio_profile == PROFILE_AUDIO_AAC &&
-							    ctx->bit_rate <= 5000000 )
-								off += sprintf(m.dlna_pn+off, "BL_L3L_SD_AAC");
-							else if( vc->level <= 31 &&
-							         audio_profile == PROFILE_AUDIO_AAC &&
-							         ctx->bit_rate <= 15000000 )
-								off += sprintf(m.dlna_pn+off, "BL_L31_HD_AAC");
-							else
-								goto mp4_mp_fallback;
-						}
-						else if( vc->width  <= 1280 &&
-						         vc->height <= 720 )
+						else
						{
-							if( vc->level <= 31 &&
-							    audio_profile == PROFILE_AUDIO_AAC &&
-							    ctx->bit_rate <= 15000000 )
-								off += sprintf(m.dlna_pn+off, "BL_L31_HD_AAC");
-							else if( vc->level <= 32 &&
-							         audio_profile == PROFILE_AUDIO_AAC &&
-							         ctx->bit_rate <= 21000000 )
-								off += sprintf(m.dlna_pn+off, "BL_L32_HD_AAC");
-							else
-								goto mp4_mp_fallback;
+							off -= 9;
+							goto mp4_mp_fallback;
						}
+					}
+					else if( vc->width  <= 720 &&
+							 vc->height <= 576 )
+					{
+						if( vc->level == 30 &&
+							audio_profile == PROFILE_AUDIO_AAC &&
+							ctx->bit_rate <= 5000000 )
+							off += sprintf(m.dlna_pn+off, "BL_L3L_SD_AAC");
+						else if( vc->level <= 31 &&
+								 audio_profile == PROFILE_AUDIO_AAC &&
+								 ctx->bit_rate <= 15000000 )
+							off += sprintf(m.dlna_pn+off, "BL_L31_HD_AAC");
+						else
+							goto mp4_mp_fallback;
+					}
+					else if( vc->width  <= 1280 &&
+							 vc->height <= 720 )
+					{
+						if( vc->level <= 31 &&
+							audio_profile == PROFILE_AUDIO_AAC &&
+							ctx->bit_rate <= 15000000 )
+							off += sprintf(m.dlna_pn+off, "BL_L31_HD_AAC");
+						else if( vc->level <= 32 &&
+								 audio_profile == PROFILE_AUDIO_AAC &&
+								 ctx->bit_rate <= 21000000 )
+							off += sprintf(m.dlna_pn+off, "BL_L32_HD_AAC");
						else
							goto mp4_mp_fallback;
-						break;
-					case FF_PROFILE_H264_MAIN:
-					mp4_mp_fallback:
-						off += sprintf(m.dlna_pn+off, "MP_");
-						/* AVC MP4 SD profiles - 10 Mbps max */
-						if( vc->width  <= 720 &&
-						    vc->height <= 576 &&
-						    vc->bit_rate <= 10000000 )
-						{
-							sprintf(m.dlna_pn+off, "SD_");
-							if( audio_profile == PROFILE_AUDIO_AC3 )
-								off += sprintf(m.dlna_pn+off, "AC3");
-							else if( audio_profile == PROFILE_AUDIO_AAC ||
-							         audio_profile == PROFILE_AUDIO_AAC_MULT5 )
-								off += sprintf(m.dlna_pn+off, "AAC_MULT5");
-							else if( audio_profile == PROFILE_AUDIO_MP3 )
-								off += sprintf(m.dlna_pn+off, "MPEG1_L3");
-							else
-								m.dlna_pn[10] = '\0';
-						}
-						else if( vc->width  <= 1280 &&
-						         vc->height <= 720 &&
-						         vc->bit_rate <= 15000000 &&
-						         audio_profile == PROFILE_AUDIO_AAC )
-						{
-							off += sprintf(m.dlna_pn+off, "HD_720p_AAC");
-						}
-						else if( vc->width  <= 1920 &&
-						         vc->height <= 1080 &&
-						         vc->bit_rate <= 21000000 &&
-						         audio_profile == PROFILE_AUDIO_AAC )
-						{
-							off += sprintf(m.dlna_pn+off, "HD_1080i_AAC");
-						}
-						if( strlen(m.dlna_pn) <= 11 )
-						{
-							DPRINTF(E_DEBUG, L_METADATA, "No DLNA profile found for %s file %s\n",
-								m.dlna_pn, basename(path));
-							free(m.dlna_pn);
-							m.dlna_pn = NULL;
-						}
-						break;
-					case FF_PROFILE_H264_HIGH:
-						if( vc->width  <= 1920 &&
-						    vc->height <= 1080 &&
-						    vc->bit_rate <= 25000000 &&
-						    audio_profile == PROFILE_AUDIO_AAC )
-						{
-							off += sprintf(m.dlna_pn+off, "HP_HD_AAC");
-						}
-						break;
-					default:
-						DPRINTF(E_DEBUG, L_METADATA, "AVC profile [%d] not recognized for file %s\n",
-							vc->profile, basename(path));
-						free(m.dlna_pn);
-						m.dlna_pn = NULL;
-						break;
					}
+					else if( vc->width  <= 1920 &&
+						vc->height <= 1080 &&
+						vc->bit_rate <= 25000000 &&
+						audio_profile == PROFILE_AUDIO_AAC )
+					{
+						off += sprintf(m.dlna_pn+off, "HP_HD_AAC");
+					}
+					else
+						goto mp4_mp_fallback;
				}
				else
				{
@@ -1311,6 +1258,47 @@ GetVideoMetadata(const char * path, char * name)
				}
				DPRINTF(E_DEBUG, L_METADATA, "Stream %d of %s is h.264\n", video_stream, basename(path));
				break;
+
+mp4_mp_fallback:
+				off += sprintf(m.dlna_pn+off, "MP_");
+				/* AVC MP4 SD profiles - 10 Mbps max */
+				if( vc->width  <= 720 &&
+					vc->height <= 576 &&
+					vc->bit_rate <= 10000000 )
+				{
+					sprintf(m.dlna_pn+off, "SD_");
+					if( audio_profile == PROFILE_AUDIO_AC3 )
+						off += sprintf(m.dlna_pn+off, "AC3");
+					else if( audio_profile == PROFILE_AUDIO_AAC ||
+							 audio_profile == PROFILE_AUDIO_AAC_MULT5 )
+						off += sprintf(m.dlna_pn+off, "AAC_MULT5");
+					else if( audio_profile == PROFILE_AUDIO_MP3 )
+						off += sprintf(m.dlna_pn+off, "MPEG1_L3");
+					else
+						m.dlna_pn[10] = '\0';
+				}
+				else if( vc->width  <= 1280 &&
+						 vc->height <= 720 &&
+						 vc->bit_rate <= 15000000 &&
+						 audio_profile == PROFILE_AUDIO_AAC )
+				{
+					off += sprintf(m.dlna_pn+off, "HD_720p_AAC");
+				}
+				else if( vc->width  <= 1920 &&
+						 vc->height <= 1080 &&
+						 vc->bit_rate <= 21000000 &&
+						 audio_profile == PROFILE_AUDIO_AAC )
+				{
+					off += sprintf(m.dlna_pn+off, "HD_1080i_AAC");
+				}
+				if( strlen(m.dlna_pn) <= 11 )
+				{
+					DPRINTF(E_DEBUG, L_METADATA, "No DLNA profile found for %s file %s\n",
+						m.dlna_pn, basename(path));
+					free(m.dlna_pn);
+					m.dlna_pn = NULL;
+				}
+				break;
			case CODEC_ID_MPEG4:
			        fourcc[0] = vc->codec_tag     & 0xff;
			        fourcc[1] = vc->codec_tag>>8  & 0xff;
--
1.7.8.3

